source/adapters/level_zero/v2/event_pool.hpp

```diff
@@ -0,0 +1,49 @@
+//===--------- event_pool.hpp - Level Zero Adapter ------------------------===//
+//
+// Copyright (C) 2024 Intel Corporation
+//
+// Part of the Unified-Runtime Project, under the Apache License v2.0 with LLVM
+// Exceptions. See LICENSE.TXT
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+#pragma once
+
+#include <memory>
+#include <mutex>
+#include <stack>
+
+#include <unordered_map>
+#include <ur/ur.hpp>
+#include <ur_api.h>
+#include <vector>
+#include <ze_api.h>
+
+#include "../device.hpp"
+#include "common.hpp"
+#include "event.hpp"
+#include "event_provider.hpp"
+
+namespace v2 {
+
+class event_pool {
+public:
+  event_pool(std::unique_ptr<event_provider> Provider)
+      : Provider(std::move(Provider)){};
+  event_pool(event_pool &&other) : Provider(std::move(other.Provider)) {}
+
+  ~event_pool(){};
+
+  DeviceId Id() { return Provider->device()->Id; };
+
+  ur_event *allocate();
+  void free(ur_event *event);
+
+private:
+  std::deque<ur_event> Events;
```

Comment:
Hm, do we really need to keep cache of ur_events as well? They are pretty cheap to construct (do you think this will change?) so why not just create a new one for each allocation and only reuse ze_event_handles?

Reply:
The existing implementation caches the event structure allocations. I think it makes sense. While a fast-path malloc is indeed very quick, the worst-case is not (mmap + page faults). Having a collection of pre-allocated (and pre-faulted) events in a contiguous array may help us avoid latency spikes.

Whether that's going to be measurable it's hard to say, but I say we keep it for now and we can reevaluate once more of the implementation is done and we can run benchmarks.